plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcConnectionCreation.java

```diff
@@ -0,0 +1,163 @@
+        assertOpenConnections("DROP TABLE copy_of_nation", 3);
+    }
+
+    @Test
+    public void testOpenConnectionsForInformationSchemas()
+    {
+        assertOpenConnections("SELECT * FROM information_schema.schemata", 1);
+        assertOpenConnections("SELECT * FROM information_schema.tables", 1);
+        assertOpenConnections("SELECT * FROM information_schema.columns", 5);
+    }
+
+    @Test
+    public void testOpenConnectionsForPushdownQueries()
+    {
+        assertOpenConnections("SELECT * FROM nation LIMIT 1", 3);
+        assertOpenConnections("SELECT * FROM nation ORDER BY nationkey LIMIT 1", 3);
+        assertOpenConnections("SELECT * FROM nation WHERE nationkey = 1", 3);
+        assertOpenConnections("SELECT avg(nationkey) FROM nation", 3);
+        assertOpenConnections("SELECT * FROM nation n JOIN region r USING(regionkey)", 6);
+    }
+
+    private void assertOpenConnections(String sql, int expected)
+    {
+        int before = openConnectionCount.get();
+        getQueryRunner().execute(sql);
+        int after = openConnectionCount.get();
+        assertThat(after - before).isLessThanOrEqualTo(expected);
+    }
+
+    private class ReusingConnectionH2Module
+            implements Module
+    {
+        @Override
+        public void configure(Binder binder) {}
+
+        @Provides
+        @Singleton
+        @ForBaseJdbc
+        public JdbcClient provideJdbcClient(BaseJdbcConfig config, ConnectionFactory connectionFactory, IdentifierMapping identifierMapping)
+        {
+            return new TestingH2JdbcClient(config, connectionFactory, identifierMapping);
+        }
+
+        @Provides
+        @Singleton
+        @ForBaseJdbc
+        public ConnectionFactory getConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
+        {
+            DriverConnectionFactory factory = new DriverConnectionFactory(new Driver(), config, credentialProvider);
+            return new ConnectionFactory()
+            {
+                @Override
+                public Connection openConnection(ConnectorSession session)
+                        throws SQLException
+                {
+                    openConnectionCount.incrementAndGet();
+                    Connection connection = factory.openConnection(session);
+                    connectionCreations.put(connection, new Exception());
+                    return new ForwardingConnection()
+                    {
+                        @Override
+                        protected Connection getDelegate()
+                        {
+                            return connection;
+                        }
+
+                        @Override
+                        public void close()
+                                throws SQLException
+                        {
+                            closedConnectionCount.incrementAndGet();
+                            super.close();
+                            connectionCreations.remove(connection);
```

Comment:
Why this specific ordering?
Do you want `connectionCreations` to not get freed when super.close() fails?
Otherwise I'd swap the order.